samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -0,0 +1,273 @@
+
+    metadataStore.delete(toStoreKey(ssp, taskName));
+  }
+
+  /**
+   * For {@link Startpoint}s keyed only by {@link SystemStreamPartition}, this method remaps the Startpoints for the specified
+   * SystemStreamPartition to SystemStreamPartition+{@link TaskName} for all tasks provided by the {@link SystemStreamPartitionGrouper}
+   * This method is not atomic or thread-safe. The intent is for the Samza Processor's coordinator to use this
+   * method to assign the Startpoints to the appropriate tasks.
+   * @param ssp The {@link SystemStreamPartition} that a {@link Startpoint} is keyed to.
+   * @param grouper The {@link SystemStreamPartitionGrouper} is used to determine what the task names are.
+   * @return The list of {@link TaskName}s
+   */
+  public Set<TaskName> groupStartpointsPerTask(SystemStreamPartition ssp, SystemStreamPartitionGrouper grouper) {
+    Preconditions.checkNotNull(metadataStore, "Underlying metadata store not available");
+    Preconditions.checkNotNull(ssp, "SystemStreamPartition cannot be null");
+    Preconditions.checkNotNull(grouper, "SystemStreamPartitionGrouper cannot be null");
+
+    Startpoint startpoint = readStartpoint(ssp);
+    HashMap<TaskName, Startpoint> result = new HashMap<>();
+
+    LOG.info("Grouping Startpoint keyed on SSP: {} to tasks determined by the grouper: {}", ssp, grouper.getClass().getCanonicalName());
+
+    // Re-map startpoints from SSP-only keys to SSP+TaskName keys
+    Map<TaskName, Set<SystemStreamPartition>> group = grouper.group(Sets.newHashSet(ssp));
+    group.forEach((taskName, sspSet) -> {
+        if (sspSet.size() != 1) {
+          throw new SamzaException(String.format("Expected only one SSP per task from grouper: %s",
+              grouper.getClass().getCanonicalName()));
+        }
+
+        SystemStreamPartition sspFromGrouper = sspSet.iterator().next();
+        if (!Objects.equals(ssp, sspFromGrouper)) {
+          throw new SamzaException(String.format("Expected SSP: %s, but got: %s from grouper: %s", ssp, sspFromGrouper,
+              grouper.getClass().getCanonicalName()));
+        }
+
+        Startpoint startpointForTask = readStartpointForTask(ssp, taskName);
+
+        // Existing startpoint keyed by SSP+TaskName takes precedence and will not be overwritten.
+        if (Objects.isNull(startpointForTask)) {
+          writeStartpointForTask(ssp, taskName, startpoint);
+          result.put(taskName, startpoint);
+          LOG.info("Startpoint for SSP: {} remapped with task: {}.", ssp, taskName);
+        } else {
+          LOG.info("Startpoint for SSP: {} and task: {} already exists and will not be overwritten.", ssp, taskName);
+        }
+      });
+
+    // Delete startpoint keyed by only this SSP
+    deleteStartpoint(ssp);
+
+    return ImmutableSet.copyOf(group.keySet());
+  }
+
+  /**
+   * Relinquish resources held by the underlying {@link MetadataStore}
+   */
+  public void stop() {
+    if (Objects.nonNull(metadataStore)) {
```

Comment:
1. Do we need to make stop idempotent?
2. Why do we need to set metadatastore as null here. Would be better to use a different boolean variable here to make the stop idempotent.

Reply:
Change of direction. Took your suggestion instead.